pkcs11/src/new/functions/session_management.rs

```diff
@@ -35,21 +36,67 @@ impl<'a> Session<'a> {
         unsafe { Rv::from(get_pkcs11!(self.client(), C_CloseSession)(self.handle())).into_result() }
     }
 
-    pub fn login(&self, user_type: UserType, pin: &str) -> Result<()> {
+    // Do not fail if the user is already logged in. It happens if another session on the same slot
+    // has already called the log in operation. Record the login call and only log out when there
+    // aren't anymore sessions requiring log in state.
+    pub fn login(&mut self, user_type: UserType, pin: &str) -> Result<()> {
```

Comment:
Hmm, I was wondering - would it be safer to save the PIN inside `Pkcs11` and use it from there? That way the user can set it once and then scrape it from memory and let this crate handle the rest. Otherwise, they'd have to keep track of where they use it, how it's handled after use etc...

Reply:
Yes I guess we could and this is somehow related to my other comment: it depends if we assume all sessions to belong to the same trusted user? As in, could session be given to non-trusted third parties for them to use? Third-parties would have to know the PIN in order to use some functionality.
Or we just say (which I think is fair enough) that all session are trusted.